File: dqops/src/main/java/com/dqops/connectors/databricks/DatabricksParametersSpec.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */

package com.dqops.connectors.databricks;

import com.dqops.connectors.ConnectionProviderSpecificParameters;
import com.dqops.core.secrets.SecretValueLookupContext;
import com.dqops.core.secrets.SecretValueProvider;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.sources.BaseProviderParametersSpec;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import picocli.CommandLine;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * Databricks connection parameters.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class DatabricksParametersSpec extends BaseProviderParametersSpec
        implements ConnectionProviderSpecificParameters {

    private static final ChildHierarchyNodeFieldMapImpl<DatabricksParametersSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(BaseProviderParametersSpec.FIELDS) {
        {
        }
    };

    @CommandLine.Option(names = {"--databricks-host"}, description = "Databricks host name")
    @JsonPropertyDescription("Databricks host name. Supports also a ${DATABRICKS_HOST} configuration with a custom environment variable.")
    private String host;

    @CommandLine.Option(names = {"--databricks-port"}, description = "Databricks port number")
    @JsonPropertyDescription("Databricks port number. The default port is 443. Supports also a ${DATABRICKS_PORT} configuration with a custom environment variable.")
    private String port;

    @CommandLine.Option(names = {"--databricks-catalog"}, description = "Databricks catalog name.")
    @JsonPropertyDescription("Databricks catalog name. Supports also a ${DATABRICKS_CATALOG} configuration with a custom environment variable.")
    private String catalog;

    @CommandLine.Option(names = {"--databricks-user"}, description = "(Obsolete) Databricks user name.")
    @JsonPropertyDescription("(Obsolete) Databricks user name. Supports also a ${DATABRICKS_USER} configuration with a custom environment variable.")
    private String user;

    @CommandLine.Option(names = {"--databricks-password"}, description = "(Obsolete) Databricks database password.")
    @JsonPropertyDescription("(Obsolete) Databricks database password. Supports also a ${DATABRICKS_PASSWORD} configuration with a custom environment variable.")
    private String password;

    @CommandLine.Option(names = {"--databricks-http-path"}, description = "Databricks http path to the warehouse. For example: /sql/1.0/warehouses/<warehouse instance id>")
    @JsonPropertyDescription("Databricks http path to the warehouse. For example: /sql/1.0/warehouses/<warehouse instance id>. Supports also a ${DATABRICKS_HTTP_PATH} configuration with a custom environment variable.")
    private String httpPath;

    @CommandLine.Option(names = {"--databricks-access-token"}, description = "Databricks access token for the warehouse.")
    @JsonPropertyDescription("Databricks access token the warehouse. Supports also a ${DATABRICKS_ACCESS_TOKEN} configuration with a custom environment variable.")
    private String accessToken;

    @CommandLine.Option(names = {"--databricks-initialization-sql"}, description = "Custom SQL that is executed after connecting to Databricks.")
    @JsonPropertyDescription("Custom SQL that is executed after connecting to Databricks.")
    private String initializationSql;

    @CommandLine.Option(names = {"-D"}, description = "Databricks additional properties that are added to the JDBC connection string")
    @JsonPropertyDescription("A dictionary of custom JDBC parameters that are added to the JDBC connection string, a key/value dictionary.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> properties;

    /**
     * Returns the host name.
     * @return Host name.
     */
    public String getHost() {
        return host;
    }

    /**
     * Sets the host name.
     * @param host New host name.
     */
    public void setHost(String host) {
        setDirtyIf(!Objects.equals(this.host, host));
        this.host = host;
    }

    /**
     * Returns the port number. The value should store an environment variable expression or a numeric postgresql port number.
     * @return Port name or an expression to be extracted.
     */
    public String getPort() {
        return port;
    }

    /**
     * Sets the port number.
     * @param port Port name.
     */
    public void setPort(String port) {
        setDirtyIf(!Objects.equals(this.port, port));
        this.port = port;
    }

    /**
     * Returns a catalog name.
     * @return Catalog name.
     */
    public String getCatalog() {
        return catalog;
    }

    /**
     * Sets a catalog name.
     * @param catalog Catalog name.
     */
    public void setCatalog(String catalog) {
        setDirtyIf(!Objects.equals(this.catalog, catalog));
        this.catalog = catalog;
    }


    /**
     * Returns the user that is used to log in to the data source (JDBC user or similar).
     * @return User name.
     */
    public String getUser() {
        return user;
    }

    /**
     * Sets a user name.
     * @param user User name.
     */
    public void setUser(String user) {
        setDirtyIf(!Objects.equals(this.user, user));
        this.user = user;
    }

    /**
     * Returns a password used to authenticate to the server.
     * @return Password.
     */
    public String getPassword() {
        return password;
    }

    /**
     * Sets a password that is used to connect to the database.
     * @param password Password.
     */
    public void setPassword(String password) {
        setDirtyIf(!Objects.equals(this.password, password));
        this.password = password;
    }

    /**
     * Returns the httpPath for warehouse.
     * @return HttpPath for warehouse.
     */
    public String getHttpPath() {
        return httpPath;
    }

    /**
     * Sets the httpPath for warehouse.
     * @param httpPath HttpPath.
     */
    public void setHttpPath(String httpPath) {
        setDirtyIf(!Objects.equals(this.httpPath, httpPath));
        this.httpPath = httpPath;
    }

    /**
     * Returns the access token for warehouse.
     * @return Access token for warehouse.
     */
    public String getAccessToken() {
        return accessToken;
    }

    /**
     * Sets the accessToken for warehouse.
     * @param accessToken accessToken for warehouse.
     */
    public void setAccessToken(String accessToken) {
        setDirtyIf(!Objects.equals(this.accessToken, accessToken));
        this.accessToken = accessToken;
    }

    /**
     * Returns an initialization SQL that is executed after opening the connection.
     * @return
     */
    public String getInitializationSql() {
        return initializationSql;
    }

    /**
     * Sets an SQL query that is executed after opening a connection.
     * @param initializationSql Initialization sql.
     */
    public void setInitializationSql(String initializationSql) {
        this.setDirtyIf(!Objects.equals(this.initializationSql, initializationSql));
        this.initializationSql = initializationSql;
    }

    /**
     * Returns a key/value map of additional properties that are included in the JDBC connection string.
     * @return Key/value dictionary of additional JDBC properties.
     */
    public Map<String, String> getProperties() {
        return properties;
    }

    /**
     * Sets a dictionary of additional connection parameters that are added to the JDBC connection string.
     * @param properties Key/value dictionary with extra parameters.
     */
    public void setProperties(Map<String, String> properties) {
        setDirtyIf(!Objects.equals(this.properties, properties));
        this.properties = properties != null ? Collections.unmodifiableMap(properties) : null;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Creates and returns a deep copy of this object.
     */
    @Override
    public DatabricksParametersSpec deepClone() {
        DatabricksParametersSpec cloned = (DatabricksParametersSpec)super.deepClone();
        return cloned;
    }

    /**
     * Creates a trimmed and expanded version of the object without unwanted properties, but with all variables like ${ENV_VAR} expanded.
     * @param secretValueProvider Secret value provider.
     * @param lookupContext Secret lookup context.
     * @return Trimmed and expanded version of this object.
     */
    public DatabricksParametersSpec expandAndTrim(SecretValueProvider secretValueProvider, SecretValueLookupContext lookupContext) {
        DatabricksParametersSpec cloned = this.deepClone();
        cloned.host = secretValueProvider.expandValue(cloned.host, lookupContext);
        cloned.port = secretValueProvider.expandValue(cloned.port, lookupContext);
        cloned.catalog = secretValueProvider.expandValue(cloned.catalog, lookupContext);
        cloned.user = secretValueProvider.expandValue(cloned.user, lookupContext);
        cloned.password = secretValueProvider.expandValue(cloned.password, lookupContext);
        cloned.httpPath = secretValueProvider.expandValue(cloned.httpPath, lookupContext);
        cloned.accessToken = secretValueProvider.expandValue(cloned.accessToken, lookupContext);
        cloned.initializationSql = secretValueProvider.expandValue(cloned.initializationSql, lookupContext);
        cloned.properties = secretValueProvider.expandProperties(cloned.properties, lookupContext);

        return cloned;
    }

    /**
     * Returns a database name.
     * @return Database name.
     */
    @Override
    public String getDatabase(){
        return catalog;
    }

}

File: distribution/python/dqops/client/models/databricks_parameters_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.databricks_parameters_spec_properties import (
        DatabricksParametersSpecProperties,
    )


T = TypeVar("T", bound="DatabricksParametersSpec")


@_attrs_define
class DatabricksParametersSpec:
    """
    Attributes:
        host (Union[Unset, str]): Databricks host name. Supports also a ${DATABRICKS_HOST} configuration with a custom
            environment variable.
        port (Union[Unset, str]): Databricks port number. The default port is 443. Supports also a ${DATABRICKS_PORT}
            configuration with a custom environment variable.
        catalog (Union[Unset, str]): Databricks catalog name. Supports also a ${DATABRICKS_CATALOG} configuration with a
            custom environment variable.
        user (Union[Unset, str]): Databricks user name. Supports also a ${DATABRICKS_USER} configuration with a custom
            environment variable.
        password (Union[Unset, str]): Databricks database password. Supports also a ${DATABRICKS_PASSWORD} configuration
            with a custom environment variable.
        http_path (Union[Unset, str]): Databricks http path to the warehouse. For example:
            /sql/1.0/warehouses/<warehouse instance id>. Supports also a ${DATABRICKS_HTTP_PATH} configuration with a custom
            environment variable.
        access_token (Union[Unset, str]): Databricks access token the warehouse. Supports also a
            ${DATABRICKS_ACCESS_TOKEN} configuration with a custom environment variable.
        properties (Union[Unset, DatabricksParametersSpecProperties]): A dictionary of custom JDBC parameters that are
            added to the JDBC connection string, a key/value dictionary.
        database (Union[Unset, str]):
    """

    host: Union[Unset, str] = UNSET
    port: Union[Unset, str] = UNSET
    catalog: Union[Unset, str] = UNSET
    user: Union[Unset, str] = UNSET
    password: Union[Unset, str] = UNSET
    http_path: Union[Unset, str] = UNSET
    access_token: Union[Unset, str] = UNSET
    properties: Union[Unset, "DatabricksParametersSpecProperties"] = UNSET
    database: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        host = self.host
        port = self.port
        catalog = self.catalog
        user = self.user
        password = self.password
        http_path = self.http_path
        access_token = self.access_token
        properties: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.properties, Unset):
            properties = self.properties.to_dict()

        database = self.database

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if host is not UNSET:
            field_dict["host"] = host
        if port is not UNSET:
            field_dict["port"] = port
        if catalog is not UNSET:
            field_dict["catalog"] = catalog
        if user is not UNSET:
            field_dict["user"] = user
        if password is not UNSET:
            field_dict["password"] = password
        if http_path is not UNSET:
            field_dict["http_path"] = http_path
        if access_token is not UNSET:
            field_dict["access_token"] = access_token
        if properties is not UNSET:
            field_dict["properties"] = properties
        if database is not UNSET:
            field_dict["database"] = database

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.databricks_parameters_spec_properties import (
            DatabricksParametersSpecProperties,
        )

        d = src_dict.copy()
        host = d.pop("host", UNSET)

        port = d.pop("port", UNSET)

        catalog = d.pop("catalog", UNSET)

        user = d.pop("user", UNSET)

        password = d.pop("password", UNSET)

        http_path = d.pop("http_path", UNSET)

        access_token = d.pop("access_token", UNSET)

        _properties = d.pop("properties", UNSET)
        properties: Union[Unset, DatabricksParametersSpecProperties]
        if isinstance(_properties, Unset):
            properties = UNSET
        else:
            properties = DatabricksParametersSpecProperties.from_dict(_properties)

        database = d.pop("database", UNSET)

        databricks_parameters_spec = cls(
            host=host,
            port=port,
            catalog=catalog,
            user=user,
            password=password,
            http_path=http_path,
            access_token=access_token,
            properties=properties,
            database=database,
        )

        databricks_parameters_spec.additional_properties = d
        return databricks_parameters_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
